rfcs/0014-opendal.md

```diff
@@ -0,0 +1,196 @@
+    .finish()?;
+```
+
+This compatibility layer enables a low-risk, incremental migration strategy. In the initial phase, complex wrappers like `CachedObjectStore` can remain unchanged and be passed directly to OpenDAL's compatibility layer. This ensures zero behavioral changes for systems that heavily depend on current caching semantics, such as ZeroFS.
+
+Over time, we can evaluate OpenDAL's built-in layers for feature parity, replace custom wrappers with OpenDAL equivalents where beneficial, and contribute missing functionality upstream to OpenDAL if needed.
+
+This approach decouples the migration timeline from the need to immediately replace all custom logic, reducing risk and allowing thorough validation at each step.
+
+<!-- TOC --><a name="analysis-of-direct-usages"></a>
+### Analysis of Direct Usages
+
+SlateDB has significant `object_store` integration across multiple components. This section analyzes the **direct call sites** of `object_store::ObjectStore` APIs in core business logic to understand the migration scope.
+
+We skip the wrapper implementations (`CachedObjectStore`, `RetryingObjectStore`) here since they will be replaced by OpenDAL's layer system. Instead, we focus on where core components directly invoke `object_store` APIs.
+
+Fortunately, direct usage of `object_store` APIs is highly concentrated in just two components: `TableStore` and `ObjectStoreSequencedStorageProtocol`. Most other components use object stores indirectly through these abstractions, which significantly reduces the migration surface area.
+
+<!-- TOC --><a name="tablestore"></a>
+#### TableStore
+
+The `TableStore` component (`tablestore.rs`) has four direct call sites to `object_store` APIs: 
+
+1. `head()` for getting object metadata in `ReadOnlyObject::len()`
+2. `get_range()` for reading byte ranges in `ReadOnlyObject::read_range()`
+3. `get()` for reading full files in `ReadOnlyObject::read()`
+4. `delete()` for SST file cleanup in `delete_sst()`
+
+The `ReadOnlyObject` struct, which implements the `ReadOnlyBlob` trait, serves as the primary interface for reading SSTable blocks, indexes, and filters. Additionally, the component uses `object_store::buffered::BufWriter` for efficient SST writes in `write_sst_in_object_store()`. List operations for WAL and SST discovery are delegated to the underlying object store via the `ObjectStores` wrapper.
+
+Migrating these four call sites to OpenDAL equivalents is straightforward, though we need to ensure OpenDAL provides equivalent buffered write capabilities and that stream-based list operations work seamlessly.
+
+<!-- TOC --><a name="objectstoresequencedstorageprotocol"></a>
+#### ObjectStoreSequencedStorageProtocol
+
+The `ObjectStoreSequencedStorageProtocol` component (`transactional_object/object_store.rs`) has three direct call sites:
+
+1. `put_opts()` with `PutMode::Create` for writing new versioned manifests with CAS semantics
+2. `get()` for reading specific manifest versions
```

Comment:
Nit: we now use `ObjectStoreSequencedStorageProtocol` for `.compactions` files as well. #1064